presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneJoinColumns.java

```diff
@@ -0,0 +1,105 @@
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.join;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.strictProject;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.values;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.testing.Closeables.closeAllRuntimeException;
+
+public class TestPruneJoinColumns
+{
+    private RuleTester tester;
+
+    @BeforeClass
+    public void setUp()
+    {
+        tester = new RuleTester();
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
+    {
+        closeAllRuntimeException(tester);
+        tester = null;
+    }
+
+    private static PlanNode buildProjectedJoin(PlanBuilder p, Predicate<Symbol> projectionFilter)
+    {
+        Symbol leftKey = p.symbol("leftKey", BIGINT);
+        Symbol leftValue = p.symbol("leftValue", BIGINT);
+        Symbol rightKey = p.symbol("rightKey", BIGINT);
+        Symbol rightValue = p.symbol("rightValue", BIGINT);
+        List<Symbol> outputs = ImmutableList.of(leftKey, leftValue, rightKey, rightValue);
+        return p.project(
+                Assignments.identity(
+                        outputs.stream()
+                                .filter(projectionFilter)
+                                .collect(toImmutableList())),
+                p.join(
+                        JoinNode.Type.INNER,
+                        p.values(leftKey, leftValue),
+                        p.values(rightKey, rightValue),
+                        ImmutableList.of(new JoinNode.EquiJoinClause(leftKey, rightKey)),
+                        outputs,
+                        Optional.empty(),
+                        Optional.empty(),
+                        Optional.empty()));
+    }
+
+    @Test
+    public void testNotAllOutputsReferenced()
+    {
+        tester.assertThat(new PruneJoinColumns())
+                .on(p -> buildProjectedJoin(p, symbol -> symbol.getName().equals("rightValue")))
+                .matches(
+                        strictProject(
+                                ImmutableMap.of("rightValue", PlanMatchPattern.expression("rightValue")),
+                                join(
+                                        JoinNode.Type.INNER,
+                                        ImmutableList.of(equiJoinClause("leftKey", "rightKey")),
+                                        Optional.empty(),
+                                        values(ImmutableList.of("leftKey", "leftValue")),
+                                        values(ImmutableList.of("rightKey", "rightValue")))
+                                        .withExactOutputs("rightValue")));
+    }
+
+    @Test
+    public void testAllInputsReferenced()
+    {
+        tester.assertThat(new PruneJoinColumns())
+                .on(p -> buildProjectedJoin(p, symbol -> true))
+                .doesNotFire();
+    }
```

Comment:
I'll add the cross join case.  Except for whether the node is a cross join, the rule never looks at the join type field, so I don't think a non-inner-join test is worthwhile.